### PR TITLE
Split bsg_set_tile_x_y variables into bsg_tile_config_vars

### PR DESCRIPTION
**Pull request by drichmond**
It turns out that archive files are amazing. This PR splits the tile configuration variables into a separate file so that the function bsg_set_tile_x_y isn't included in CUDA binaries IF linking uses a .a file.

Not ready to merge - still testing with replicant.

### CODE REVIEW

**Comment by drichmond** [comment]
Halp, CI failed

**Comment by drichmond** [comment]
Passes regression in replicant

File: software/bsg_manycore_lib/bsg_tile_config_vars.c [hunk 1]
```diff
@@ -0,0 +1,14 @@
+// these are private variables
+// we do not make them volatile
+// so that they may be cached
+
+int __bsg_x = -1;
```
Comment (vb000) [inline software/bsg_manycore_lib/bsg_tile_config_vars.c]: What do you think about adding `__attribute__ ((section (".dmem")))`?

Comment (drichmond) [reply software/bsg_manycore_lib/bsg_tile_config_vars.c]: Seems fine to me?

Comment (drichmond) [reply software/bsg_manycore_lib/bsg_tile_config_vars.c]: Like this? 

`__attribute__ ((section (".dmem"))) int __bsg_x = -1;`

Comment (drichmond) [reply software/bsg_manycore_lib/bsg_tile_config_vars.c]: Fixed! 

And added comments

**Comment by drichmond** [comment]
@bornaehsani Are my comments accurate?

File: software/bsg_manycore_lib/bsg_tile_config_vars.c (hunk 1 above)
Comment (drichmond) [reply software/bsg_manycore_lib/bsg_tile_config_vars.c]: Hmm, with the .dmem attributes it fails regression on bsg_print_stat

Comment (drichmond) [reply software/bsg_manycore_lib/bsg_tile_config_vars.c]: @vb000 Did you mean .sdata.dmem?

Comment (drichmond) [reply software/bsg_manycore_lib/bsg_tile_config_vars.c]: When I remove the attributes, all tests pass. When I add them in (using .dmem) it fails.

Comment (drichmond) [reply software/bsg_manycore_lib/bsg_tile_config_vars.c]: Without attributes: 

```
Disassembly of section .sdata.dmem:

00001060 <__bsg_tile_group_id-0x4>:
    1060:       0000                    unimp
    1062:       424c                    lw      x11,4(x12)

00001064 <__bsg_tile_group_id>:
    1064:       ffff                    0xffff
    1066:       ffff                    0xffff

00001068 <__bsg_tile_group_id_y>:
    1068:       ffff                    0xffff
    106a:       ffff                    0xffff

0000106c <__bsg_tile_group_id_x>:
    106c:       ffff                    0xffff
    106e:       ffff                    0xffff

00001070 <__bsg_grid_dim_y>:
    1070:       ffff                    0xffff
    1072:       ffff                    0xffff

00001074 <__bsg_grid_dim_x>:
    1074:       ffff                    0xffff
    1076:       ffff                    0xffff

00001078 <__bsg_grp_org_y>:
    1078:       ffff                    0xffff
    107a:       ffff                    0xffff

0000107c <__bsg_grp_org_x>:
    107c:       ffff                    0xffff
    107e:       ffff                    0xffff

00001080 <__bsg_id>:
    1080:       ffff                    0xffff
    1082:       ffff                    0xffff

00001084 <__bsg_y>:
    1084:       ffff                    0xffff
    1086:       ffff                    0xffff

00001088 <__bsg_x>:
    1088:       ffff                    0xffff
    108a:       ffff                    0xffff
```

Comment (drichmond) [reply software/bsg_manycore_lib/bsg_tile_config_vars.c]: With attributes
```
Disassembly of section .dmem:

00001060 <__bsg_tile_group_id>:
    1060:       ffff                    0xffff
    1062:       ffff                    0xffff

00001064 <__bsg_tile_group_id_y>:
    1064:       ffff                    0xffff
    1066:       ffff                    0xffff

00001068 <__bsg_tile_group_id_x>:
    1068:       ffff                    0xffff
    106a:       ffff                    0xffff

0000106c <__bsg_grid_dim_y>:
    106c:       ffff                    0xffff
    106e:       ffff                    0xffff

00001070 <__bsg_grid_dim_x>:
    1070:       ffff                    0xffff
    1072:       ffff                    0xffff

00001074 <__bsg_grp_org_y>:
    1074:       ffff                    0xffff
    1076:       ffff                    0xffff

00001078 <__bsg_grp_org_x>:
    1078:       ffff                    0xffff
    107a:       ffff                    0xffff

0000107c <__bsg_id>:
    107c:       ffff                    0xffff
    107e:       ffff                    0xffff

00001080 <__bsg_y>:
    1080:       ffff                    0xffff
    1082:       ffff                    0xffff

00001084 <__bsg_x>:
    1084:       ffff                    0xffff
    1086:       ffff                    0xffff

Disassembly of section .sdata.dmem:

00001088 <_bsg_data_end_addr-0x4>:
    1088:       0000                    unimp
    108a:       424c                    lw      x11,4(x12)
```

Comment (vb000) [reply software/bsg_manycore_lib/bsg_tile_config_vars.c]: I think there's something fishy going on with `__bsg_tile_group_id` here. If you're using this as a part of `.a` file, linker script automatically puts these in one of the ".dmem" sections. So, we don't need this attribute. Is that the case?

I'll try to add these attributes in the linker script refactor. You could merge these without them...

File: software/spmd/bsg_cuda_lite_runtime/barrier/Makefile [hunk 2]
```diff
@@ -20,7 +20,7 @@ all: main.run
 
 KERNEL_NAME ?=kernel_barrier
 
-OBJECT_FILES=main.o bsg_set_tile_x_y.o bsg_printf.o kernel_barrier.o
+OBJECT_FILES=main.o bsg_tile_config_vars.o bsg_set_tile_x_y.o bsg_printf.o kernel_barrier.o
```
Comment (tommydcjung) [inline software/spmd/bsg_cuda_lite_runtime/barrier/Makefile]: maybe a separate list for common objs and reuse them?

File: software/bsg_manycore_lib/bsg_tile_config_vars.c (hunk 1 above)
Comment (drichmond) [reply software/bsg_manycore_lib/bsg_tile_config_vars.c]: These are part of the .a file, I don't think I've changed anything?

Regardless, I'll leave the attributes out for now.

File: software/spmd/bsg_cuda_lite_runtime/barrier/Makefile (hunk 2 above)
Comment (drichmond) [reply software/spmd/bsg_cuda_lite_runtime/barrier/Makefile]: Ideally, we'd re-use the .a file that is generated. But that's not part of this PR. Perhaps @vb000 can comment, or add it in the linker script refactor?

Comment (drichmond) [reply software/spmd/bsg_cuda_lite_runtime/barrier/Makefile]: (i.e, yes I agree with you :p )

Comment (vb000) [reply software/spmd/bsg_cuda_lite_runtime/barrier/Makefile]: Right, I totally missed that! We use a make variable called `$(SPMD_COMMON_OBJECTS)` that does that. Sure, I could do that.

Comment (drichmond) [reply software/spmd/bsg_cuda_lite_runtime/barrier/Makefile]: Why not link with the .a file, though? This works in baseline...

Comment (drichmond) [reply software/spmd/bsg_cuda_lite_runtime/barrier/Makefile]: And we do it in SPMD

Comment (vb000) [reply software/spmd/bsg_cuda_lite_runtime/barrier/Makefile]: We have `.a` as well. SPMD_COMMON_OBJECTS was the old way of doing it. I think `.a` should supersede it.

Comment (drichmond) [reply software/spmd/bsg_cuda_lite_runtime/barrier/Makefile]: Perfect! This all makes sense now

Comment (drichmond) [reply software/spmd/bsg_cuda_lite_runtime/barrier/Makefile]: Filed an issue

**Comment by tommydcjung** [comment]
https://gitlab.com/dpetrisko/bsg_manycore/-/jobs/474982756

these jobs might not have finished correctly

**Comment by drichmond** [comment]
> https://gitlab.com/dpetrisko/bsg_manycore/-/jobs/474982756
> 
> these jobs might not have finished correctly

The ones "Without BSG_FINISH"?

**Comment by drichmond** [comment]
I fixed the CUDA Makefiles as well. We no longer use individual .o files for bsg_manycore_lib objects. This passes cosimulation regression on F1.

**Comment by tommydcjung** [comment]
I would rebase and make sure that CI passes

**Pull request merged**